rfc/rfc-83/rfc-83.md

```diff
@@ -0,0 +1,209 @@
+After the change, now `HoodieIncrSource` would behave like this:
+1. Use the last checkpoint and query configs to build `IncrementalQueryAnalyzer` and get `QueryContext`.
+2. Run snapshot/incremental query with the info from `QueryContext` and this will trigger the underlying `IncrementalRelation`.
+3. Filter the returned data with plugins like `SnapshotLoadQuerySplitter`, drop metadata fields, report metrics, etc.
+
+### Refactor Cloud Events Incremental Source
+#### Incremental Logic Abstraction
+Now let's move on to cloud events incremental source for S3/GCS. Due to its vastly different logic, 
+it's very challenging to plug in `IncrementalQueryAnalyzer` and migrate logic we have for `HoodieIncrSource` to 
+cloud events incremental source. Even if we managed to do so, the code would not be maintainable. Hence, a refactoring is necessary.
+
+We've already covered the updated logic flow of `HoodieIncrSource` and cloud events incremental source.
+Comparison of these sources reveals three abstract steps:
+- Get metadata
+- Fetch data
+- Post actions
+
+The following is a table to describe what `HoodieIncrSource` and cloud events incremental source need to do for each step.
+
+| *                        | Get Metadata                                                                                                                                                                               | Fetch Data                                                                         | Post Actions                                         |
+|--------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------|------------------------------------------------------|
+| `HoodieIncrSource`       | - Pass checkpoint and query config to `IncrementalQueryAnalyzer` and get `QueryContext`                                                                                                    | - Use `QueryContext` and `IncrementalRelation` to query data                       | - Filter the result<br/>- Report metrics<br/> - etc. |
+| Cloud Events Incr Source | - Pass checkpoint and query config to `IncrementalQueryAnalyzer` and get `QueryContext`<br/>- Fetch eligible `CloudObjectMetadata` from cloud metadata Hudi table<br/>- Filter the results | - Use `QueryContext` and `CloudObjectMetadata` to fetch objects from cloud storage | - Report metrics<br/>- etc.                          |
+
+Note that Flink's `IncrementalInputSplits` can also be abstracted into these flow, but it's not in the scope of this RFC.
+
+#### IncrementalQueryRunner
+With the abstraction above, we can come up with the interface `IncrementalQueryRunner`.
+```java
+public interface IncrementalQueryRunner<M, D> {
+  
+  M getMetadata();
+
+  D fetchData();
```

Comment:
It still looks like a mini-batch style interfaces, assuming the dataset can be returned in one shot, can the interface be refactored as per-row or streaming style?

Reply:
I think the `D` here can be a single row, it mainly depends on the implementation. If we are talking about buffering then I think we will need to add more plugins like `IncrementalQueryBuffer` and put it into the implementation of `IncrementalQueryRunner`. 

Do you have any examples/code code I can refer to?

Reply:
The streaming operators are usually chained together for multiple functions, they also have some interfaces to define the dataset end barrier. I kind of think maybe mini-batch is a compromise for both batch and streaming in this case, just need to think through the pre/post hanling of each mini-batch.